GH-37857: [Python][Dataset] Expose file size to python dataset

eeroel opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change

Allow passing known file sizes to `make_fragment`, to avoid potential network requests.

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

### What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

### Are these changes tested?

Yes, tests with S3 that file size gets used.

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

### Are there any user-facing changes?

Yes, new function arguments.

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->
* Closes: #37857

mapleFU commented:
Would you mind change `GS` to `GH`?

mapleFU commented:
So this `size` is expect to avoid a `HEAD` or `LIST` request in Object Store? Seems that this might be a bit dangerous if the size doesn't match

eeroel commented:
> So this `size` is expect to avoid a `HEAD` or `LIST` request in Object Store? Seems that this might be a bit dangerous if the size doesn't match

Yes, that's the idea. In Delta Lake and I believe in Iceberg as well, the objects are not changed after they are written, so the size that those formats store in their metadata should be correct always.

mapleFU commented:
Thats a good idea, but I guess when we create InputStream in Object Store, the internal of stream will also issuing an HEAD request ( https://github.com/apache/arrow/blob/main/cpp/src/arrow/filesystem/s3fs.cc#L1232 ), I wonder if these are all duplicates...

eeroel commented:
> Thats a good idea, but I guess when we create InputStream in Object Store, the internal of stream will also issuing an HEAD request ( https://github.com/apache/arrow/blob/main/cpp/src/arrow/filesystem/s3fs.cc#L1232 ), I wonder if these are all duplicates...

When the source has file size it will be passed here: `https://github.com/apache/arrow/blob/ebc23687cb0376e24a0f002fe710db5ad891c674/cpp/src/arrow/filesystem/s3fs.cc#L2546`, so the function will early exit (https://github.com/apache/arrow/blob/ebc23687cb0376e24a0f002fe710db5ad891c674/cpp/src/arrow/filesystem/s3fs.cc#L1224)

But that request you linked is also problematic with regards to threading. I wonder if it would be difficult to refactor it to use `SubmitIO` ?

mapleFU commented:
Can we use this part of interface: https://github.com/apache/arrow/commit/4f8504f28877995ed179e35b077e6df6c14cac9c ?

By the way, `-1` is a bit misleading here...

eeroel commented:
> Can we use this part of interface: [4f8504f](https://github.com/apache/arrow/commit/4f8504f28877995ed179e35b077e6df6c14cac9c) ?

Which part do you mean? In the current PR FileSource is constructed with FileInfos and FileInfos get constructed with size: https://github.com/apache/arrow/pull/37868/files#diff-82a2d9a8424dfb436344c0802dc36275cd0fb87e2e9eb6399745fba43799f7e5R63

> By the way, `-1` is a bit misleading here...
Yes, agreed it's not nice. I couldn't get the extension to compile when defaulting to size=None, got some errors about `Python object "cannot be passed as a varargs parameter"`. I'm not familiar with Cython so I don't really know what that's about...

mapleFU commented:
Ok, I mean can we use the one with `CustomOpen`, but this change also lgtm. You can firstly fix the ci, I'll ask the committers for review.

eeroel commented:
> Ok, I mean can we use the one with `CustomOpen`, but this change also lgtm. You can firstly fix the ci, I'll ask the committers for review.

Hmm, I wonder why this breaks 🤔 It works on my machine... https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/48125445#L3684

eeroel commented:
Noticed that the code was always passing the size to FileSource even if it was the default -1... Surprisingly test passed but it probably worked due to some internal validation in FileSource/FileInfo. Fixed now. But if someone has an idea of how to better handle the optional int value, I'm happy to change

westonpace commented:
> But that request you linked is also problematic with regards to threading

Can you expand on this?

> By the way, -1 is a bit misleading here...

There is a `kNoSize` constant in `filesystem.h` that could be used in a few places but I don't know how much clearer that makes things.  We are on C++17 now so we can use https://en.cppreference.com/w/cpp/utility/optional but that would be a considerably more significant refactor.  I think -1 / kNoSize is probably ok for now.

westonpace reviewed (changes requested):
Let's get CI passing and then this change seems ok to me.  @jorisvandenbossche any thoughts on the python API changes (though they are pretty minor, just adding a new optional parameter to `make_fragment`)?

Might be interesting to @wjones127 as well.

eeroel commented:
> > But that request you linked is also problematic with regards to threading
> 
> Can you expand on this?
> 
I was running some experiments looking at the debug logs, and it seems that these HEAD requests always get executed on the main (?) thread. And from the code it also seems that way, it's not Async. So when a dataset consists of multiple fragments the file reads start effectively in sequence, and the latency from each HEAD adds up.

wjones127 commented on python/pyarrow/_dataset_parquet.pyx:
```diff
@@ -226,7 +226,7 @@ cdef class ParquetFileFormat(FileFormat):
 
     def make_fragment(self, file, filesystem=None,
-                      Expression partition_expression=None, row_groups=None):
+                      Expression partition_expression=None, row_groups=None, size=-1):
```
I think you'll want to do something like `max_chunksize` from this code:

https://github.com/apache/arrow/blob/aca1d3eeed3775c2f02e9f5d59d62478267950b1/python/pyarrow/_flight.pyx#L1132

eeroel replied:
Thanks, updated to something like this, so now only the `cdef` functions get -1

eeroel commented:
Updated the size to default to None instead, and fixed pre-commit linter errors so hopefully caught all

westonpace commented:
> EDIT: I believe it's originating from here 

Ah, yes, I think you are right.  We'd probably want to add an `OpenAsync` to `FileSource`.

eeroel commented:
> > EDIT: I believe it's originating from here
> 
> Ah, yes, I think you are right. We'd probably want to add an `OpenAsync` to `FileSource`.

I can open an issue and a PR with an attempted solution

wjones127 reviewed (approved):
This looks good. Thanks @eeroel .

The CI failures look unrelated.

I'll given Weston a day or two in case he has comments and otherwise I will merge.

mapleFU reviewed (approved):
Also LGTM

mapleFU commented:
@wjones127 Can we just try to merge this? Or should waiting for other committer?

mapleFU commented:
@eeroel Would you mind rebase this patch? Since some CI failure has been fixed now.

eeroel commented:
> @eeroel Would you mind rebase this patch? Since some CI failure has been fixed now.

Done

eeroel commented:
Errors don't seem related?

mapleFU commented:
@pitrou @jorisvandenbossche @bkietz 

Would you mind review this patch? I think the C++ part look good to me. But I'm not so familiar with Python part.

pitrou commented on python/pyarrow/tests/test_dataset.py:
```diff
@@ -981,6 +981,48 @@ def test_make_fragment(multisourcefs):
 
+@pytest.mark.parquet
+@pytest.mark.s3
+def test_make_fragment_with_size(s3_example_simple):
```
We don't need S3 for these tests, do we? Can we simply use the local filesystem, or even a mock filesystem?

eeroel replied:
Good point, I was focused on an S3 use case so didn't consider this. I tried to convert this to use the same mock filesystem from `multisourcefs` as the above test, but in that case the exceptions do not get raised, and the `to_table()` calls return incomplete data of 24 rows (in both size=1 and size=1000000). Need to look into this more to see if it's a bug on the Python side or the C++ side.

eeroel replied:
Looked into this further, the incomplete data issue was just a bug in the test so a false alarm. But basically this test can't use local or mock filesystems because they don't currently implement `OpenInputFile` that takes `FileInfo`, so the size argument just gets ignored. Given this, do you think it would be OK to just use S3 here, as that is one concrete implementation in which this optimization is used?

pitrou replied:
@eeroel Sorry for the delay. Yes, this sounds ok to me. Can you add a comment explaining this?

pitrou reviewed (changes requested):
Thanks for this.
Can you add a test that this succeeds when the right size is passed? Perhaps on the C++ side, instead of only in Python?

jorisvandenbossche commented on python/pyarrow/_dataset.pyx:
```diff
@@ -1230,15 +1233,16 @@ cdef class FileFormat(_Weakrefable):
     def make_fragment(self, file, filesystem=None,
-                      Expression partition_expression=None):
+                      Expression partition_expression=None,
+                      size=None):
```
Could maybe also name this `file_size` to be more explicit? (eg it's not the "size" of the fragment in memory, for example in terms of rows)

eeroel replied:
Thanks, made these changes

eeroel commented:
> Thanks for this. Can you add a test that this succeeds when the right size is passed? Perhaps on the C++ side, instead of only in Python?

Added a test on the Python side but where do you think would be a good place for the C++ test? There are already these tests from the original implementation, but do you mean we should also write something that tests the path from FileSource creation onwards? https://github.com/apache/arrow/commit/4f8504f28877995ed179e35b077e6df6c14cac9c#diff-5748697423aff1b6f4688662b451378684a3ba0e1d3b13a8eca267925e3afb39

eeroel commented on cpp/src/arrow/filesystem/filesystem.h:
```diff
@@ -60,6 +60,8 @@ struct ARROW_EXPORT FileInfo : public util::EqualityComparable<FileInfo> {
   explicit FileInfo(std::string path, FileType type = FileType::Unknown)
       : path_(std::move(path)), type_(type) {}
+  explicit FileInfo(std::string path, int64_t size, FileType type = FileType::Unknown)
+      : path_(std::move(path)), type_(type), size_(size) {}
```
I think this is not relevant for this PR now that I removed the extra FileSource constructor?

eeroel commented on cpp/src/arrow/dataset/file_base.h:
```diff
@@ -54,7 +54,11 @@ class ARROW_DS_EXPORT FileSource : public util::EqualityComparable<FileSource> {
         filesystem_(std::move(filesystem)),
         compression_(compression) {}
-
+  FileSource(std::string path, int64_t size, std::shared_ptr<fs::FileSystem> filesystem,
```
Thanks, I changed this to use FileInfo from Python side now

mapleFU commented:
Would you mind fix the lint here?

eeroel commented:
> Would you mind fix the lint here?
> 
> ```
> INFO:archery:Running Docker linter
> --- original//arrow/python/pyarrow/tests/test_dataset.py
> +++ fixed//arrow/python/pyarrow/tests/test_dataset.py
> @@ -1003,7 +1003,7 @@
>      fragments_with_size = [file_format.make_fragment(path, fs, file_size=size)
>                             for path, size in zip(paths, sizes_true)]
>      dataset_with_size = ds.FileSystemDataset(
> -            fragments_with_size, format=file_format, schema=table.schema, filesystem=fs
> +        fragments_with_size, format=file_format, schema=table.schema, filesystem=fs
>      )
>      tbl = dataset.to_table()
>      assert tbl.equals(table)
> ```

Fixed this, also removed a stray "some_file" that I had accidentally committed

mapleFU commented:
@jorisvandenbossche @pitrou would you mind take a look again?

mapleFU commented:
Gently ping @pitrou

pitrou commented on python/pyarrow/tests/test_dataset.py:
```diff
@@ -981,6 +981,59 @@ def test_make_fragment(multisourcefs):
+    )
+
+    # invalid range
+    with pytest.raises(OSError, match='HTTP status 416'):
```
I'm not sure the exact HTTP error is relevant here? Though we can probably keep it until it's not reliable anymore :-)

eeroel replied:
Yeah maybe it's not ideal but at least "range not satisfiable" is the expected outcome, instead of some error upstream for example? Since probably there's no way to validate the size before making the request

pitrou reviewed (changes requested):
Sorry for the delay in reviewing this. Here are some suggestions.

eeroel commented on python/pyarrow/_dataset.pyx:
```diff
@@ -96,27 +96,33 @@ def _get_parquet_symbol(name):
-        c_source = CFileSource(move(c_path), move(c_filesystem))
 
+        if file_size >= 0:
+            c_size = file_size
+            info = FileInfo(c_path, size=c_size)
+            c_info = info.unwrap()
```
Hmm tried this but got an error that CFileInfo expects one argument. Left it as is for now.

eeroel commented on cpp/src/arrow/dataset/file_base.h:
```diff
@@ -54,7 +54,6 @@ class ARROW_DS_EXPORT FileSource : public util::EqualityComparable<FileSource> {
       : file_info_(std::move(path)),
         filesystem_(std::move(filesystem)),
         compression_(compression) {}
-
```
Fixed these

eeroel commented:
Thanks for comments @pitrou, I think they are now addressed

eeroel commented on python/pyarrow/_dataset.pyx:
```diff
@@ -96,27 +96,33 @@ def _get_parquet_symbol(name):
 
 
-cdef CFileSource _make_file_source(object file, FileSystem filesystem=None):
+cdef CFileSource _make_file_source(object file, FileSystem filesystem=None, int64_t file_size=-1):
```
> Instead of coercing file_size here and also in make_fragment, could we instead allow this function to accept None and move it all here?

Hmm this doesn't compile for me

eeroel replied:
> Instead of coercing file_size here and also in make_fragment, could we instead allow this function to accept None and move it all here?

I tried that initially but couldn't get it to work, hence the awkward integer argument. Sadly the earlier conversation is lost because I have force-pushed since that :(

eeroel replied:
> Also, based on call sites below it looks like you mean to require kwargs for filesystem and file_size

Hmm this doesn't compile for me (* is not allowed)

bkietz replied:
Huh, I didn't know that was disallowed for cdef functions... maybe it makes parsing pointer args harder?

Re the default arg, I made this metaPR https://github.com/eeroel/arrow/pull/1 which compiled and passed your test case on my machine

eeroel replied:
Thanks a lot, much cleaner now! Can't for the life of me remember why I couldn't get it to work earlier 🤔

mapleFU commented:
Ping @pitrou

pitrou commented:
Thank you for contributing @eeroel !

eeroel commented:
Thanks everyone for the help, and @mapleFU in particular for pushing this forward!

ldacey commented:
Should this fail with bad file_size values? I tried changing the file_size to -10000 and it still succeed for me. I am not sure if GCS uses this information though (I am using fsspec for the filesystem which uses the gcsfs library). I think I am using it wrong because I cannot get it to fail in general, regardless of whether the size I input is the real size of the file.

I didn't see a fragment.size attribute or anything to check other than fragment.metadata.serialized_size which is different 
```python
fragment = file_format.make_fragment(path, filesystem=dataset.filesystem, partition_expression=expression, file_size=-10000)
print(fragment)

<pyarrow.dataset.ParquetFileFragment path=bucket/discard/year=2023/part-0.parquet partition=[year=2023]>
<pyarrow.dataset.ParquetFileFragment path=bucket/discard/year=2024/part-0.parquet partition=[year=2024]>

# fresh dataset written with version 15.0
<pyarrow._parquet.FileMetaData object at 0x7f725816ecf0>
  created_by: parquet-cpp-arrow version 15.0.0
  num_columns: 56
  num_rows: 21420
  num_row_groups: 5
  format_version: 2.6
  serialized_size: 34164

```

eeroel commented:
> Should this fail with bad file_size values? I tried changing the file_size to -10000 and it still succeed for me. I am not sure if GCS uses this information though (I am using fsspec for the filesystem which uses the gcsfs library). I think I am using it wrong because I cannot get it to fail in general, regardless of whether the size I input is the real size of the file.

Did you also try to create a dataset with those fragments and read it? There's no validation when the fragments are constructed, but it should fail when the parquet reader tries to start reading the file, in here: https://github.com/apache/arrow/blob/21ffd82c05c93b873ae3c27128eb8604ed0c735f/cpp/src/parquet/file_reader.cc#L476. It would make sense to handle zero and negative sizes on the Python side though...

Regarding fsspec, the file size information will only get used for Arrow internal file system implementations, and I believe currently it's only used for S3.

mapleFU commented:
> Should this fail with bad file_size values? I tried changing the file_size to -10000 and it still succeed for me.

I guess if fileSize is invalid, parsing a parquet footer might failed when size is bad. In other format the behavior might be undefined.

ldacey commented:
Here are some more details:

```python
ic(pa.__version__)
ic(fs)
dataset = ds.dataset(gcs_path, filesystem=fs)
ic(dataset.filesystem)

table = dataset.to_table()
ic(table.num_rows)

file_format = ds.ParquetFileFormat()
paths = dataset.files

original_fragments = [frag for frag in dataset.get_fragments()]
original_dataset = ds.FileSystemDataset(
    original_fragments, format=file_format, schema=table.schema, filesystem=dataset.filesystem
)
assert dataset.to_table().equals(original_dataset.to_table())
ic(dataset.to_table().equals(original_dataset.to_table()))

wrong_fragments = [] 
for path in dataset.files:
    fake_size = 55555555555
    actual_size = dataset.filesystem.get_file_info(path).size
    ic(actual_size, fake_size)
    fragment = file_format.make_fragment(path, filesystem=dataset.filesystem, file_size=fake_size)
    wrong_fragments.append(fragment)
    
test = ds.FileSystemDataset(
    wrong_fragments, format=file_format, schema=table.schema, filesystem=dataset.filesystem
)

assert dataset.to_table().equals(test.to_table())
ic(dataset.to_table().equals(test.to_table()))

___

ic| pa.__version__: '15.0.0'
ic| fs: <gcsfs.core.GCSFileSystem object at 0x7f13ea37dd90>
ic| dataset.filesystem: <pyarrow._fs.PyFileSystem object at 0x7f13b5b2d130>
ic| table.num_rows: 23491
ic| dataset.to_table().equals(original_dataset.to_table()): True
ic| actual_size: 4237841, fake_size: 55555555555
ic| dataset.to_table().equals(test.to_table()): True

```

I was using dataset.filesystem which is a pyarrow filesystem (generated from the gcsfs filesystem). In that example, I made the file size some random large number and you can see it does not match the real size yet the table constructs and still matches the original table.  I thought that would fail, unless I am doing something wrong. I have pyarrow 15 installed though.

eeroel commented:
> Here are some more details:
> 
> ```python
> ic(pa.__version__)
> ic(fs)
> dataset = ds.dataset(gcs_path, filesystem=fs)
> ic(dataset.filesystem)
> 
> table = dataset.to_table()
> ic(table.num_rows)
> 
> file_format = ds.ParquetFileFormat()
> paths = dataset.files
> 
> original_fragments = [frag for frag in dataset.get_fragments()]
> original_dataset = ds.FileSystemDataset(
>     original_fragments, format=file_format, schema=table.schema, filesystem=dataset.filesystem
> )
> assert dataset.to_table().equals(original_dataset.to_table())
> ic(dataset.to_table().equals(original_dataset.to_table()))
> 
> wrong_fragments = [] 
> for path in dataset.files:
>     fake_size = 55555555555
>     actual_size = dataset.filesystem.get_file_info(path).size
>     ic(actual_size, fake_size)
>     fragment = file_format.make_fragment(path, filesystem=dataset.filesystem, file_size=fake_size)
>     wrong_fragments.append(fragment)
>     
> test = ds.FileSystemDataset(
>     wrong_fragments, format=file_format, schema=table.schema, filesystem=dataset.filesystem
> )
> 
> assert dataset.to_table().equals(test.to_table())
> ic(dataset.to_table().equals(test.to_table()))
> 
> ___
> 
> ic| pa.__version__: '15.0.0'
> ic| fs: <gcsfs.core.GCSFileSystem object at 0x7f13ea37dd90>
> ic| dataset.filesystem: <pyarrow._fs.PyFileSystem object at 0x7f13b5b2d130>
> ic| table.num_rows: 23491
> ic| dataset.to_table().equals(original_dataset.to_table()): True
> ic| actual_size: 4237841, fake_size: 55555555555
> ic| dataset.to_table().equals(test.to_table()): True
> ```
> 
> I was using dataset.filesystem which is a pyarrow filesystem (generated from the gcsfs filesystem). In that example, I made the file size some random large number and you can see it does not match the real size yet the table constructs and still matches the original table. I thought that would fail, unless I am doing something wrong. I have pyarrow 15 installed though.

Do you also get this result if you set `fake_size` to a negative number? With a too-large size this is expected, because the size is not actually used in this case (it's only used with Arrow S3fs implementation at the moment).

ldacey commented:
Yes, I changed fake_size to -9999 and reran it and it still worked. But since I am not using S3 (only have access to GCS and ADLSgen2) perhaps it is just ignored entirely.

```python
ic| pa.__version__: '15.0.0'
ic| fs: <gcsfs.core.GCSFileSystem object at 0x7f55e8b88a10>
ic| dataset.filesystem: <pyarrow._fs.PyFileSystem object at 0x7f562f099230>
ic| table.num_rows: 23491
ic| dataset.to_table().equals(original_dataset.to_table()): True
ic| actual_size: 4237841, fake_size: -9999
ic| dataset.to_table().equals(test.to_table()): True
```

My original plan was to take a look at deltalake (delta-rs library) which already uses `make_fragments()`. Since the transaction log (`get_add_actions()`) has the actual file sizes then we could pass these to `make_fragments()` for some potential efficiency correct?

```python
        if not filesystem or pyarrow.__version__ >= "15.0":
            file_sizes = self.get_add_actions().to_pydict()
            file_sizes = {
                x: y for x, y in zip(file_sizes["path"], file_sizes["size_bytes"])
            }

        format = ParquetFileFormat(
            read_options=parquet_read_options,
            default_fragment_scan_options=ParquetFragmentScanOptions(pre_buffer=True),
        )

        fragments = []
        for file, part_expression in self._table.dataset_partitions(
            self.schema().to_pyarrow(), partitions
        ):
            if pyarrow.__version__ >= "15.0":
                fragment = format.make_fragment(
                    file,
                    filesystem=filesystem,
                    partition_expression=part_expression,
                    file_size=file_sizes[file],
                )
            else:
                fragment = format.make_fragment(
                    file,
                    filesystem=filesystem,
                    partition_expression=part_expression,
                )
            fragments.append(fragment)
```

eeroel commented:
> Yes, I changed fake_size to -9999 and reran it and it still worked. But since I am not using S3 (only have access to GCS and ADLSgen2) perhaps it is just ignored entirely.

OK, thanks for confirming. I think it's good to check this, I didn't add a test case for negative values for this PR. Could be a bug, or it's possible there's some validation somewhere along the chain that silently ignores the value.

> My original plan was to take a look at deltalake (delta-rs library) which already uses make_fragments(). Since the transaction log (get_add_actions()) has the actual file sizes then we could pass these to make_fragments() for some potential efficiency correct?

Yep, that was the motivation for this PR! I actually implemented this in deltalake for the filesystem implementation, so if you can use that then the optimization should already be applied: https://github.com/delta-io/delta-rs/blob/1b6c830aae4553d2a079a2bf42b024863fcbbb40/python/deltalake/table.py#L1035

To use this with the Arrow GCS implementation, I think the OpenFile method should be updated to check the size from FileInfo.
